File: 47_Permutations_II.java
class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        ArrayList<List<Integer>> result = new ArrayList<>();

        /* Sanity check */
        if(nums.length == 0)
            return result;

        Arrays.sort(nums);
        dfs(nums, result, 0);

        return result;
    }

    private void dfs(int[] nums, List<List<Integer>> result, int index) {

        if (index == nums.length - 1) {
            ArrayList<Integer> result_part = new ArrayList<>();
            for(int a : nums)
                result_part.add(new Integer(a));
            result.add(result_part);
            return;
        }

        dfs(nums, result, index + 1);
        HashSet<Integer> used = new HashSet<>();
        used.add(Integer.valueOf(nums[index]));
        for(int i = index + 1; i < nums.length; ++i) {
            if(used.contains(Integer.valueOf(nums[i])))
                continue;
            swap(nums, index, i);
            dfs(nums, result, index + 1);
            swap(nums, index, i);
            used.add(nums[i]);
        }
    }

    private void swap(int[] nums, int a, int b) {
        int c = nums[a];
        nums[a] = nums[b];
        nums[b] = c;
    }

}

File: 47_Permutations_II.cpp
class Solution {
 private:
  vector<vector<int>> result;

  inline void swap(vector<int>& a, int i, int j) {
    if (i == j) return;
    a[i] ^= a[j];
    a[j] ^= a[i];
    a[i] ^= a[j];
  }

  inline void dfs(vector<int>& nums, int current) {
    if (current == nums.size() - 1)
      result.push_back(nums);
    else {
      set<int> checked;
      dfs(nums, current + 1);
      checked.insert(nums[current]);
      for (int j = current + 1; j < nums.size(); ++j) {
        if (checked.find(nums[j]) != checked.end()) continue;
        swap(nums, current, j);
        dfs(nums, current + 1);
        swap(nums, current, j);
        checked.insert(nums[j]);
      }
    }
  }

 public:
  vector<vector<int>> permuteUnique(vector<int>& nums) {
    dfs(nums, 0);
    return result;
  }
};